File: backend/app/services/compliance.py
from sqlalchemy import case, func, literal

from app.models.media_file_scan import MediaFileScan
from app.models.quality_profile import QualityProfile


COMPLIANCE_CHECK_COUNT = 4


def build_media_scan_compliant_piece_count(active_profile: QualityProfile | None):
    codec_condition = literal(False)
    pixel_format_condition = literal(False)
    file_format_condition = literal(False)

    if active_profile is not None:
        normalized_codec = active_profile.codec.strip().lower()
        codec_condition = func.lower(func.coalesce(MediaFileScan.codec, "")) == normalized_codec

        if active_profile.pixel_format:
            allowed_pixel_formats = [
                value.strip().lower()
                for value in active_profile.pixel_format.split(",")
                if value.strip()
            ]
            if allowed_pixel_formats:
                pixel_format_condition = func.lower(func.coalesce(MediaFileScan.pixel_format, "")).in_(
                    allowed_pixel_formats
                )

        if active_profile.file_format:
            allowed_file_formats = [
                value.strip().lower().lstrip(".")
                for value in active_profile.file_format.split(",")
                if value.strip()
            ]
            if allowed_file_formats:
                file_format_condition = func.lower(func.coalesce(MediaFileScan.extension, "")).in_(
                    allowed_file_formats
                )

    tag_condition = MediaFileScan.tag_status == "tag_match"

    return (
        case((codec_condition, 1), else_=0)
        + case((pixel_format_condition, 1), else_=0)
        + case((file_format_condition, 1), else_=0)
        + case((tag_condition, 1), else_=0)
    )

File: frontend/src/lib/compliance.ts
import { type MediaFileScanResult, type QualityProfile } from './api'

export type ComplianceStatus = 'compliant' | 'partial_compliant' | 'non_compliant'

const COMPLIANCE_CHECK_COUNT = 4

function normalizeCsvList(value: string | null | undefined, stripDot = false) {
  return (value ?? '')
    .split(',')
    .map((item) => item.trim().toLowerCase())
    .map((item) => (stripDot ? item.replace(/^\./, '') : item))
    .filter((item) => item.length > 0)
}

export function isCodecCompliant(result: MediaFileScanResult, profile?: QualityProfile) {
  const expectedCodec = profile?.codec?.trim().toLowerCase()
  const actualCodec = result.codec?.trim().toLowerCase()
  if (!expectedCodec || !actualCodec) {
    return false
  }
  return actualCodec === expectedCodec
}

export function isPixelFormatCompliant(result: MediaFileScanResult, profile?: QualityProfile) {
  const allowedPixelFormats = normalizeCsvList(profile?.pixel_format)
  const actualPixelFormat = result.pixel_format?.trim().toLowerCase()
  if (allowedPixelFormats.length === 0 || !actualPixelFormat) {
    return false
  }
  return allowedPixelFormats.includes(actualPixelFormat)
}

export function isFileFormatCompliant(result: MediaFileScanResult, profile?: QualityProfile) {
  const allowedFileFormats = normalizeCsvList(profile?.file_format, true)
  const actualExtension = result.extension?.trim().toLowerCase().replace(/^\./, '')
  if (allowedFileFormats.length === 0 || !actualExtension) {
    return false
  }
  return allowedFileFormats.includes(actualExtension)
}

export function isTagCompliant(result: MediaFileScanResult) {
  return result.tag_status === 'tag_match'
}

export function getCompliantPieceCount(result: MediaFileScanResult, profile?: QualityProfile) {
  return [
    isCodecCompliant(result, profile),
    isPixelFormatCompliant(result, profile),
    isFileFormatCompliant(result, profile),
    isTagCompliant(result),
  ].filter(Boolean).length
}

export function getComplianceStatus(result: MediaFileScanResult, profile?: QualityProfile): ComplianceStatus {
  const compliantPieceCount = getCompliantPieceCount(result, profile)

  if (compliantPieceCount === COMPLIANCE_CHECK_COUNT) {
    return 'compliant'
  }
  if (compliantPieceCount > 0) {
    return 'partial_compliant'
  }
  return 'non_compliant'
}
